### PR TITLE
Fix the NullArgumentException when uid is empty and improve resolve x…

### PR DESCRIPTION
**Pull request by vicancy**
…ref message

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs [hunk 1]
```diff
@@ -168,33 +174,50 @@ internal ManifestItem Transform(InternalManifestItem item)
...
+            {
+                return;
             }
 
-            if (missingXRefDetails.Count > 0)
+            var invalidXRefs = (from ae in aes
+                                      from ie in ae.InnerExceptions
+                                      select ie as InvalidCrossReferenceException into ce
+                                      where ce != null
+                                      select ce.XRefDetails).Distinct().ToList();
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: >Distinct [](start = 61, length = 8)

XRefDetails not override gethashcode yet?

File: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs [hunk 2]
```diff
@@ -168,33 +174,50 @@ internal ManifestItem Transform(InternalManifestItem item)
...
-                    // For each source file, print the first 10 invalid cross reference
-                    if (group.Count() >= maxCountPerFile)
-                    {
-                        details = $"{string.Join(", ", group.Take(maxCountPerFile).Select(i => $"\"{i.RawSource}\" in line {i.SourceStartLineNumber.ToString()}"))}, etc. Fisrt {maxCountPerFile} invalid cross reference are shown above.";
-                    }
-                    else
-                    {
-                        details = string.Join(", ", group.Select(i => $"\"{i.RawSource}\" in line {i.SourceStartLineNumber.ToString()}"));
-                    }
+            var distinctUids = invalidXRefs.Select(i => i.RawSource).Distinct().Select(s => $"\"{HttpUtility.HtmlDecode(s)}\"").ToList();
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: @jsntcy After this change, need op encode the message in html report generator?

**Comment by vwxyzh** [comment]
>                 throw new AggregateException(xrefExceptions);

why not return xrefExceptions?

---
Refers to: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs:364 in aabab41. [](commit_id = aabab412079312a6d401b115aaa5c1ad657c3740, deletion_comment = False)

File: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs (hunk 2 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: Nice catch, synced with @fenxu , will integrate with them when our change is in.

**Comment by vicancy** [comment]
>                 throw new AggregateException(xrefExceptions);

return xrefDetails instead

---
In reply to: [321126656](https://github.com/dotnet/docfx/pull/1935#issuecomment-321126656) [](ancestors = 321126656)

---
Refers to: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs:364 in aabab41. [](commit_id = aabab412079312a6d401b115aaa5c1ad657c3740, deletion_comment = False)

**Comment by hellosnow** [comment]
>         public void ProcessMrefWithNotInvalidCrossReferenceShouldWarn()

help to remove the Not in case name...

---
Refers to: test/Microsoft.DocAsCode.Build.ManagedReference.Tests/ManagedReferenceDocumentProcessorTest.cs:225 in 84a9b26. [](commit_id = 84a9b264d82b5f5d1c5b6a73f3e0677131d5f7f9, deletion_comment = False)

File: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs [hunk 3]
```diff
@@ -168,33 +171,38 @@ internal ManifestItem Transform(InternalManifestItem item)
...
+                var message = $"Details for {prefix}invalid cross reference(s): {details.ToDelimitedString(", ")}";
 
-                    Logger.LogInfo($"Invalid cross reference details: {details}", file: group.Key ?? item.LocalPathFromRoot);
+                if (group.Key != null)
+                {
+                    Logger.LogInfo(message, file: group.Key);
+                }
+                else
+                {
+                    Logger.LogInfo(message);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: >message [](start = 35, length = 7)

if raw source is null, need we keep the logic to log with item.LocalPathFromRoot?

Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: yes, that's set in FileScope outside

**Comment by hellosnow** [comment]
:shipit:

**Comment by vwxyzh** [comment]
:shipit:

**Pull request merged**